### ISSUE TITLE
What spec does method property follow

### ISSUE THREAD

**Issue by felixfbecker** [Microsoft/language-server-protocol#65]
The JSON RPC spec says that the `method` should be the name of the method to call on the target. But the protocol makes use of names like `textDocument/didOpen`. A slash is typically not a valid character for a method name, so I imply that it has some special meaning, eg call the `didOpen` method on the `textDocument` property. Initially, it looked like a [JSON Pointer](https://tools.ietf.org/html/rfc6901) to me, but a JSON Pointer always has to start with a slash:

> ```
> The ABNF syntax of a JSON Pointer is:
> 
>       json-pointer    = *( "/" reference-token )
>       reference-token = *( unescaped / escaped )
>       unescaped       = %x00-2E / %x30-7D / %x7F-10FFFF
>          ; %x2F ('/') and %x7E ('~') are excluded from 'unescaped'
>       escaped         = "~" ( "0" / "1" )
>         ; representing '~' and '/', respectively
> ```

So it should be `/textDocument/didOpen`.
I had to write a whole new JSON RPC library that allows this kind of method names. I would have much rather liked to just open a PR at one of the existing libraries to implement JSON Pointer support.

What was the design decision behind using a slash? What is it's meaning?

**Comment by dbaeumer** [comment]
We use the / to group method into logical areas. The first part denotes the group the second part the method in that group.

To my interpretation the method name is specified as a string with only the limitation that a string starting with 'rpc.' has a special meaning in the protocol. I somehow assumed that the JSON RPC libraries don't do any automatic binding to real program code on the server since JSON RPC doesn't make any assumption about the language used. Otherwise the spec would need to define a method name to for example forbid that a name can start with a number (not valid JS identifier).

JSON pointers are to my knowledge something different. The allow you to reference a JSON object inside a JSON structure. They wouldn't help with grouping methods except that they would also use '/' as a separator. 

Which library did you use for the JSON RPC communication. May be the library could offer an indirection where a handler is registered. If not handler is provided then the library could use introspection to see if the 'server' offers a 'method' with the same name.

Given all this I would opt to leave the naming as is. Changing it will be a major breakage.

**Comment by felixfbecker** [comment]
I just wanted to know the thought process behind this. Pretty much all JSON RPC libraries I saw provided a Server implementation that dispatches request to method calls on an object. The JSON RPC protocol of course makes no assumptions here, but in reality that is how most implementations work - you do a _remote procedure call_ , eg call a method on the server with given params. I understand now that you see it more as "namespaces", but that doesn't map very nice to object-oriented programming languages. 

In the end I rolled my own library and decided to path `/` methods to methods of object properties:
https://github.com/felixfbecker/php-advanced-json-rpc#nested-targets
which is exactly how a JSON Pointer works, just that it doesn't have a slash in the beginning.
